documents/design/cone_of_shame.md

```diff
@@ -0,0 +1,55 @@
+# Project Title
+
+*Issue #Number*
+571
+
+**Author:**
+- Jiajun Mao
+
+## The Problem
+
+In last year's IGVC competition, Jessie turned 180 degrees before starting traveling on the course. This is presumably caused by the current navigation model ignoring everything behind Jessie and default that nothing exists there. This caused the problem with the path planning algorithm thinking that the shortest path is to walk backward to the place where "no line exists".
+
+## Proposed Solution
+
+- The problem could potentially be solved by creating a cone behind the robot and make the robot think that that is a part of the line defining the course. Thus the robot would not travel backwards.
+- How to measure the location of the cone
+    - Subscribe to **/move_base_flex/local_costmap/costmap** and select the point on the line that is closest to the robot
```

Comment:
Since this is going to be in a `costmap_2d::Layer`, you don't actually need to subscribe to `/move_base_flex/local_costmap/costmap` to get access to the `line_layer` (it's a bit roundabout). The `costmap_2d::layer` base class has a `layered_costmap_` member, which in turn contains a `getPlugins` member function returning `std::vector<boost::shared_ptr<Layer> >*`. You should be able to check if the the `boost::shared_ptr<Layer>` is of type `LineLayer` by using `boost::dynamic_ptr_cast`, after which you can just work with the `LineLayer`'s internal map representation directly. The benefit of this is that you don't need to serialize + deserialize the map (expensive), and that there will only be lines now as opposed to lines and barrels.